File: streamlit/lib/capacity.py
"""School/sports capacity helpers."""

from __future__ import annotations

import json
from datetime import datetime
from typing import Any

from lib.db import get_conn


def get_capacity_settings() -> dict[str, Any]:
    row = get_conn().execute(
        """
        SELECT available_start_hour, available_end_hour, blocked_json
        FROM settings WHERE id = 1
        """
    ).fetchone()
    blocked = []
    try:
        blocked = json.loads(row["blocked_json"] or "[]")
    except Exception:
        blocked = []
    return {
        "available_start_hour": row["available_start_hour"],
        "available_end_hour": row["available_end_hour"],
        "blocked": blocked,
    }


def save_capacity_settings(
    start: int | None = None,
    end: int | None = None,
    blocked: list[dict[str, Any]] | None = None,
) -> None:
    cur = get_capacity_settings()
    start = start if start is not None else cur["available_start_hour"]
    end = end if end is not None else cur["available_end_hour"]
    blocked = blocked if blocked is not None else cur["blocked"]
    if end <= start:
        end = min(22, start + 4)
    get_conn().execute(
        """
        UPDATE settings SET available_start_hour = ?, available_end_hour = ?,
        blocked_json = ? WHERE id = 1
        """,
        (start, end, json.dumps(blocked)),
    )
    get_conn().commit()
    try:
        from lib.calendar_sync import mark_calendar_dirty

        mark_calendar_dirty()
    except Exception:
        pass


def fmt_hour(h: int) -> str:
    ampm = "pm" if h >= 12 else "am"
    hr = 12 if h % 12 == 0 else h % 12
    return f"{hr}{ampm}"


def available_minutes_for_date(
    dt: datetime, settings: dict[str, Any] | None = None
) -> dict[str, Any]:
    settings = settings or get_capacity_settings()
    start = settings["available_start_hour"]
    end = settings["available_end_hour"]
    window = max(0, (end - start) * 60)
    dow = dt.weekday()  # Mon=0…Sun=6 in Python
    # Convert to JS-style Sun=0…Sat=6 used in stored blocks
    js_dow = (dow + 1) % 7

    blocked_mins = 0
    labels: list[str] = []
    for slot in settings["blocked"]:
        if int(slot.get("dow", -1)) != js_dow:
            continue
        o_start = max(start, int(slot["startHour"]))
        o_end = min(end, int(slot["endHour"]))
        if o_end > o_start:
            blocked_mins += (o_end - o_start) * 60
            labels.append(
                f"{slot.get('label', 'Blocked')} ({fmt_hour(int(slot['startHour']))}–{fmt_hour(int(slot['endHour']))})"
            )

    return {
        "available_minutes": max(0, window - blocked_mins),
        "blocked_minutes": blocked_mins,
        "window_minutes": window,
        "blocked_labels": labels,
    }


def capacity_snapshot(scheduled_minutes: int, dt: datetime | None = None) -> dict[str, Any]:
    dt = dt or datetime.now()
    settings = get_capacity_settings()
    day = available_minutes_for_date(dt, settings)
    avail = day["available_minutes"]
    over = scheduled_minutes > avail
    ratio = (
        min(1.5, scheduled_minutes / avail)
        if avail > 0
        else (1.5 if scheduled_minutes else 0)
    )
    return {
        **day,
        "settings": settings,
        "scheduled_minutes": scheduled_minutes,
        "remaining_minutes": avail - scheduled_minutes,
        "overbooked": over,
        "fill_percent": round(min(100, ratio * 100)),
        "work_window_label": f"{fmt_hour(settings['available_start_hour'])}–{fmt_hour(settings['available_end_hour'])}",
    }

File: src/lib/capacity.ts
/**
 * School/sports capacity — how many mowing minutes Miles actually has in a day.
 * Work window minus blocked blocks (practice, games, dinner with parents, etc.).
 */

import { getDb } from "./db";
import { type BlockedSlot, fmtHour } from "./capacity-shared";

export type { BlockedSlot } from "./capacity-shared";
export { dowName, fmtHour } from "./capacity-shared";

export type CapacitySettings = {
  availableStartHour: number;
  availableEndHour: number;
  blocked: BlockedSlot[];
};

export function ensureCapacityColumns() {
  const db = getDb();
  const cols = db.prepare(`PRAGMA table_info(settings)`).all() as Array<{
    name: string;
  }>;
  if (!cols.some((c) => c.name === "available_start_hour")) {
    db.exec(
      `ALTER TABLE settings ADD COLUMN available_start_hour INTEGER NOT NULL DEFAULT 8`
    );
  }
  if (!cols.some((c) => c.name === "available_end_hour")) {
    db.exec(
      `ALTER TABLE settings ADD COLUMN available_end_hour INTEGER NOT NULL DEFAULT 18`
    );
  }
  if (!cols.some((c) => c.name === "blocked_json")) {
    // Seed a sample Tue/Thu after-school block — Miles can edit/clear.
    db.exec(
      `ALTER TABLE settings ADD COLUMN blocked_json TEXT NOT NULL DEFAULT '[{"dow":2,"startHour":15,"endHour":19,"label":"After school / practice"},{"dow":4,"startHour":15,"endHour":19,"label":"After school / practice"}]'`
    );
  }
}

export function getCapacitySettings(): CapacitySettings {
  ensureCapacityColumns();
  const row = getDb()
    .prepare(
      `SELECT available_start_hour, available_end_hour, blocked_json
       FROM settings WHERE id = 1`
    )
    .get() as {
    available_start_hour: number;
    available_end_hour: number;
    blocked_json: string;
  };

  let blocked: BlockedSlot[] = [];
  try {
    const parsed = JSON.parse(row.blocked_json || "[]") as unknown;
    if (Array.isArray(parsed)) {
      blocked = parsed
        .filter(
          (b): b is BlockedSlot =>
            !!b &&
            typeof b === "object" &&
            typeof (b as BlockedSlot).dow === "number" &&
            typeof (b as BlockedSlot).startHour === "number" &&
            typeof (b as BlockedSlot).endHour === "number"
        )
        .map((b) => ({
          dow: b.dow,
          startHour: b.startHour,
          endHour: b.endHour,
          label: String(b.label || "Blocked").slice(0, 60),
        }));
    }
  } catch {
    blocked = [];
  }

  return {
    availableStartHour: row.available_start_hour,
    availableEndHour: row.available_end_hour,
    blocked,
  };
}

export function saveCapacitySettings(input: {
  availableStartHour?: number;
  availableEndHour?: number;
  blocked?: BlockedSlot[];
}): CapacitySettings {
  ensureCapacityColumns();
  const current = getCapacitySettings();
  const next: CapacitySettings = {
    availableStartHour:
      input.availableStartHour ?? current.availableStartHour,
    availableEndHour: input.availableEndHour ?? current.availableEndHour,
    blocked: input.blocked ?? current.blocked,
  };

  // Guard: end after start
  if (next.availableEndHour <= next.availableStartHour) {
    next.availableEndHour = Math.min(22, next.availableStartHour + 4);
  }

  getDb()
    .prepare(
      `UPDATE settings SET
        available_start_hour = ?,
        available_end_hour = ?,
        blocked_json = ?
       WHERE id = 1`
    )
    .run(
      next.availableStartHour,
      next.availableEndHour,
      JSON.stringify(next.blocked)
    );

  return next;
}

/**
 * Available mowing minutes for a calendar date (local).
 * Overlaps blocked slots with the work window and subtracts them.
 */
export function availableMinutesForDate(
  date: Date,
  settings: CapacitySettings = getCapacitySettings()
): {
  availableMinutes: number;
  blockedMinutes: number;
  windowMinutes: number;
  blockedLabels: string[];
} {
  const dow = date.getDay();
  const start = settings.availableStartHour;
  const end = settings.availableEndHour;
  const windowMinutes = Math.max(0, (end - start) * 60);

  let blockedMinutes = 0;
  const blockedLabels: string[] = [];

  for (const slot of settings.blocked) {
    if (slot.dow !== dow) continue;
    const overlapStart = Math.max(start, slot.startHour);
    const overlapEnd = Math.min(end, slot.endHour);
    if (overlapEnd > overlapStart) {
      blockedMinutes += (overlapEnd - overlapStart) * 60;
      blockedLabels.push(
        `${slot.label} (${fmtHour(slot.startHour)}–${fmtHour(slot.endHour)})`
      );
    }
  }

  return {
    availableMinutes: Math.max(0, windowMinutes - blockedMinutes),
    blockedMinutes,
    windowMinutes,
    blockedLabels,
  };
}

/** Capacity snapshot for Today UI. */
export function capacitySnapshot(scheduledMinutes: number, date = new Date()) {
  const settings = getCapacitySettings();
  const day = availableMinutesForDate(date, settings);
  const overbooked = scheduledMinutes > day.availableMinutes;
  const ratio =
    day.availableMinutes > 0
      ? Math.min(1.5, scheduledMinutes / day.availableMinutes)
      : scheduledMinutes > 0
        ? 1.5
        : 0;

  return {
    ...day,
    settings,
    scheduledMinutes,
    remainingMinutes: day.availableMinutes - scheduledMinutes,
    overbooked,
    fillPercent: Math.round(Math.min(100, ratio * 100)),
    workWindowLabel: `${fmtHour(settings.availableStartHour)}–${fmtHour(settings.availableEndHour)}`,
  };
}
